mono/mini/interp/interp.c

```diff
@@ -3579,100 +3665,113 @@ interp_exec_method_full (InterpFrame *frame, ThreadContext *context, FrameClause
+				*sp = *u.child_frame.retval;
 				sp++;
 			}
 			MINT_IN_BREAK;
 		}
 		MINT_IN_CASE(MINT_CALL_VARARG) {
-			int num_varargs = 0;
+
 			MonoMethodSignature *csig;
 
 			frame->ip = ip;
 
-			child_frame.imethod = (InterpMethod*)frame->imethod->data_items [* (guint16 *)(ip + 1)];
+			u.child_frame.imethod = (InterpMethod*)frame->imethod->data_items [* (guint16 *)(ip + 1)];
 			/* The real signature for vararg calls */
 			csig = (MonoMethodSignature*) frame->imethod->data_items [* (guint16*) (ip + 2)];
 			/* Push all vararg arguments from normal sp to vt_sp together with the signature */
-			num_varargs = csig->param_count - csig->sentinelpos;
 			vt_sp = copy_varargs_vtstack (csig, sp, vt_sp);
 
+			csig = (MonoMethodSignature*) frame->imethod->data_items [*(guint16*)(ip + 2)];  // Refetch to conserve stack.
+			int const num_varargs = csig->param_count - csig->sentinelpos;
+
 			ip += 3;
 			sp->data.p = vt_sp;
-			child_frame.retval = sp;
+			u.child_frame.retval = sp;
 
 			/* decrement by the actual number of args */
-			sp -= child_frame.imethod->param_count + child_frame.imethod->hasthis + num_varargs;
-			child_frame.stack_args = sp;
+			sp -= u.child_frame.imethod->param_count + u.child_frame.imethod->hasthis + num_varargs;
+			u.child_frame.stack_args = sp;
 
-			interp_exec_method (&child_frame, context, error);
+			u.child_frame.parent = frame;
+			interp_exec_method (&u.child_frame, context, error);
 
 			CHECK_RESUME_STATE (context);
 
+			csig = (MonoMethodSignature*) frame->imethod->data_items [*(guint16*)(ip - 3 + 2)]; // Refetch to conserve stack.
+
 			if (csig->ret->type != MONO_TYPE_VOID) {
-				*sp = *child_frame.retval;
+				*sp = *u.child_frame.retval;
 				sp++;
 			}
 			MINT_IN_BREAK;
 		}
+
 		MINT_IN_CASE(MINT_CALL)
 		MINT_IN_CASE(MINT_VCALL)
 		MINT_IN_CASE(MINT_CALLVIRT)
 		MINT_IN_CASE(MINT_VCALLVIRT) {
-
 			frame->ip = ip;
 
-			child_frame.imethod = (InterpMethod*)frame->imethod->data_items [* (guint16 *)(ip + 1)];
+			u.child_frame.imethod = (InterpMethod*)frame->imethod->data_items [* (guint16 *)(ip + 1)];
 			ip += 2;
 			sp->data.p = vt_sp;
-			child_frame.retval = sp;
+			u.child_frame.retval = sp;
 
 			/* decrement by the actual number of args */
-			sp -= child_frame.imethod->param_count + child_frame.imethod->hasthis;
-			child_frame.stack_args = sp;
+			sp -= u.child_frame.imethod->param_count + u.child_frame.imethod->hasthis;
+			u.child_frame.stack_args = sp;
 
 			const gboolean is_virtual = ip [-2] == MINT_CALLVIRT || ip [-2] == MINT_VCALLVIRT;
 			if (is_virtual) {
 				MonoObject *this_arg = (MonoObject*)sp->data.p;
 
-				child_frame.imethod = get_virtual_method (child_frame.imethod, this_arg->vtable);
-				if (m_class_is_valuetype (this_arg->vtable->klass) && m_class_is_valuetype (child_frame.imethod->method->klass)) {
+				u.child_frame.imethod = get_virtual_method (u.child_frame.imethod, this_arg->vtable);
+
+				this_arg = (MonoObject*)sp->data.p; // Refetch to conserve stack.
```

Comment:
This could probably also use the union. If things can be reordered well enough.
Maybe too gnarly.